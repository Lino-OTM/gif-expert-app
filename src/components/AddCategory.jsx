import { useState } from "react";
import { PropTypes } from "prop-types";

export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState("");

  const onInputChange = (event) => {
    setInputValue(event.target.value);
  };

  //Perfectamente se puede desestructurar el target para ahorrar escribir event.

  const onSubmit = (event) => {
    event.preventDefault();
    //Esta funcion previene el refresh del navegador
    if (inputValue.trim().length <= 1) return;
    //Esta funcion evita mandar espacios vacios o un solo caracter.
    // setCategories((categories) => [inputValue, ...categories]);
    onNewCategory(inputValue.trim());
    setInputValue("");
    // Al enviar algo se limpia el input
  };

  return (
    <form onSubmit={onSubmit} aria-label="form">
      <input
        type="text"
        placeholder="Buscar gifs"
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,
};
