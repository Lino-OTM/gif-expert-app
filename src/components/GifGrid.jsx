import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifItem } from "./GifItem";
import { PropTypes } from "prop-types";
// import { getGifs } from "../helpers/getGifs";

export const GifGrid = ({ category }) => {
  const { images, isLoading } = useFetchGifs(category);

  return (
    <>
      <h4>{category}</h4>
      {isLoading && <h4> Cargando...</h4>}
      <div className="card-grid">
        {images.map((image) => (
          <GifItem key={image.id} {...image} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}
/*
{ ...image}
El operador spread me permite extraer las propiedades sin tener que llamarlas una por una ( title={image.title}, etc etc)
MUY UTIL SI TENGO VARIAS PROPIEDADES */
