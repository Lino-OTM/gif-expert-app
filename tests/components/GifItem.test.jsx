import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

describe("Pruebas en el componente <GifItem />", () => {
  const title = "Sparda";
  const url = "https://devil-may-cry-sparda.jpg/";

  test("Debe de coincidir con la snapshot", () => {
    const { container } = render(<GifItem title={title} url={url} />);
    expect(container).toMatchSnapshot();
  });

  test("Debe de mostrar la imagen con el URL y el ALT indicado", () => {
    render(<GifItem title={title} url={url} />);
    // Para evitar escribir varios expect( screen.getByRole("img").src o .alt o etc etc).toBe(url o title); se hace una desestructuracion de screen.getByRole("img"); y se obtiene lo siguiente:
    const { src, alt } = screen.getByRole("img");
    expect(src).toBe(url);
    expect(alt).toBe(alt);
  });

  test("Debe de mostrar el titulo en el componente", () => {
    render(<GifItem title={title} url={url} />);
    expect(screen.getByText(title)).toBeTruthy();
  });
});

// La funcion render viene de react testing library para poder crear el snapshot, en este caso se crea del GifItem.

// Si no se pone los parametros que se indicaron con los propTypes en GifItem.jsx saltara un error en consola.

// Como el titulo y el url lo puedo llegar a utilizar varias veces, es buena idea crear una constante fuera del test para poder acceder a ellas en varios lados.

// Screen.debug() sirve para ver que se esta renderizando, se debe importar de React-testing-library
