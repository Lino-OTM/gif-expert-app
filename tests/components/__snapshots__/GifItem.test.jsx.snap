// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifItem /> Debe de coincidir con la snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Sparda"
      src="https://devil-may-cry-sparda.jpg/"
    />
    <p>
      Sparda
    </p>
  </div>
</div>
`;
